Haddock 2.23.0 regression: --hyperlinked-source breaks vault-0.3.1.3

The [`vault-0.3.1.3`](http://hackage.haskell.org/package/vault-0.3.1.3) library no longer works with `stack haddock` on a GHC 8.8.1-based resolver (such as [this one](https://www.stackage.org/nightly-2019-09-27)). After some digging, I discovered that the culprit is `stack haddock`'s use of the `--hyperlinked-source` flag. To reproduce the issue, pick one of the following options (depending on your favorite build tool):

* `stack`

  <details>

  ```
  $ stack unpack vault-0.3.1.3 && cd vault-0.3.1.3
  Unpacked vault (from Hackage) to /home/rgscott/Documents/Hacking/Haskell/vault-0.3.1.3/

  $ stack --resolver nightly-2019-09-27 init
  Looking for .cabal or package.yaml files to use to init the project.
  Using cabal packages:
  - ./
  
  Selected resolver: nightly-2019-09-27
  Initialising configuration using resolver: nightly-2019-09-27
  Total number of user packages considered: 1
  Writing configuration to file: stack.yaml
  All done.

  $ stack haddock
  Stack has not been tested with GHC versions above 8.6, and using 8.8.1, this may fail
  Stack has not been tested with Cabal versions above 2.4, but version 3.0.0.0 was found, this may fail
  vault> configure (lib)
  Configuring vault-0.3.1.3...
  vault> build (lib)
  Preprocessing library for vault-0.3.1.3..
  Building library for vault-0.3.1.3..
  [1 of 5] Compiling Data.Unique.Really
  
  /home/rgscott/Documents/Hacking/Haskell/vault-0.3.1.3/src/Data/Unique/Really.hs:6:1: warning: [-Wunused-imports]
      The import of ‘Control.Applicative’ is redundant
        except perhaps to import instances from ‘Control.Applicative’
      To import instances alone, use: import Control.Applicative()
    |
  6 | import Control.Applicative ((<$>))
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  [2 of 5] Compiling Data.Vault.ST.Lazy
  
  /home/rgscott/Documents/Hacking/Haskell/vault-0.3.1.3/src/Data/Vault/ST/ST.h:15:1: warning: [-Wunused-imports]
      The import of ‘Data.Monoid’ is redundant
        except perhaps to import instances from ‘Data.Monoid’
      To import instances alone, use: import Data.Monoid()
     |
  15 | import Data.Monoid (Monoid(..))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  /home/rgscott/Documents/Hacking/Haskell/vault-0.3.1.3/src/Data/Vault/ST/ST.h:16:1: warning: [-Wunused-imports]
      The import of ‘Data.Semigroup’ is redundant
        except perhaps to import instances from ‘Data.Semigroup’
      To import instances alone, use: import Data.Semigroup()
     |
  16 | import Data.Semigroup (Semigroup((<>)))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  /home/rgscott/Documents/Hacking/Haskell/vault-0.3.1.3/src/Data/Vault/ST/ST.h:18:1: warning: [-Wunused-imports]
      The import of ‘Control.Applicative’ is redundant
        except perhaps to import instances from ‘Control.Applicative’
      To import instances alone, use: import Control.Applicative()
     |
  18 | import Control.Applicative ((<$>))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  [3 of 5] Compiling Data.Vault.Lazy
  [4 of 5] Compiling Data.Vault.ST.Strict
  
  /home/rgscott/Documents/Hacking/Haskell/vault-0.3.1.3/src/Data/Vault/ST/ST.h:15:1: warning: [-Wunused-imports]
      The import of ‘Data.Monoid’ is redundant
        except perhaps to import instances from ‘Data.Monoid’
      To import instances alone, use: import Data.Monoid()
     |
  15 | import Data.Monoid (Monoid(..))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  /home/rgscott/Documents/Hacking/Haskell/vault-0.3.1.3/src/Data/Vault/ST/ST.h:16:1: warning: [-Wunused-imports]
      The import of ‘Data.Semigroup’ is redundant
        except perhaps to import instances from ‘Data.Semigroup’
      To import instances alone, use: import Data.Semigroup()
     |
  16 | import Data.Semigroup (Semigroup((<>)))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  /home/rgscott/Documents/Hacking/Haskell/vault-0.3.1.3/src/Data/Vault/ST/ST.h:18:1: warning: [-Wunused-imports]
      The import of ‘Control.Applicative’ is redundant
        except perhaps to import instances from ‘Control.Applicative’
      To import instances alone, use: import Control.Applicative()
     |
  18 | import Control.Applicative ((<$>))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  [5 of 5] Compiling Data.Vault.Strict
  
  vault> haddock
  Preprocessing library for vault-0.3.1.3..
  Running Haddock on library for vault-0.3.1.3..
  
  src/Data/Unique/Really.hs:6:1: warning: [-Wunused-imports]
      The import of ‘Control.Applicative’ is redundant
        except perhaps to import instances from ‘Control.Applicative’
      To import instances alone, use: import Control.Applicative()
    |
  6 | import Control.Applicative ((<$>))
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  src/Data/Vault/ST/ST.h:15:1: warning: [-Wunused-imports]
      The import of ‘Data.Monoid’ is redundant
        except perhaps to import instances from ‘Data.Monoid’
      To import instances alone, use: import Data.Monoid()
     |
  15 | import Data.Monoid (Monoid(..))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  src/Data/Vault/ST/ST.h:16:1: warning: [-Wunused-imports]
      The import of ‘Data.Semigroup’ is redundant
        except perhaps to import instances from ‘Data.Semigroup’
      To import instances alone, use: import Data.Semigroup()
     |
  16 | import Data.Semigroup (Semigroup((<>)))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  src/Data/Vault/ST/ST.h:18:1: warning: [-Wunused-imports]
      The import of ‘Control.Applicative’ is redundant
        except perhaps to import instances from ‘Control.Applicative’
      To import instances alone, use: import Control.Applicative()
     |
  18 | import Control.Applicative ((<$>))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  src/Data/Vault/ST/ST.h:15:1: warning: [-Wunused-imports]
      The import of ‘Data.Monoid’ is redundant
        except perhaps to import instances from ‘Data.Monoid’
      To import instances alone, use: import Data.Monoid()
     |
  15 | import Data.Monoid (Monoid(..))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  src/Data/Vault/ST/ST.h:16:1: warning: [-Wunused-imports]
      The import of ‘Data.Semigroup’ is redundant
        except perhaps to import instances from ‘Data.Semigroup’
      To import instances alone, use: import Data.Semigroup()
     |
  16 | import Data.Semigroup (Semigroup((<>)))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  src/Data/Vault/ST/ST.h:18:1: warning: [-Wunused-imports]
      The import of ‘Control.Applicative’ is redundant
        except perhaps to import instances from ‘Control.Applicative’
      To import instances alone, use: import Control.Applicative()
     |
  18 | import Control.Applicative ((<$>))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  Haddock coverage:
   100% (  4 /  4) in 'Data.Unique.Really'
    80% ( 12 / 15) in 'Data.Vault.ST.Lazy'
    Missing documentation for:
      Vault (src/Data/Vault/ST/backends/GHC.h:18)
      Key (src/Data/Vault/ST/backends/GHC.h:19)
      Locker (src/Data/Vault/ST/backends/GHC.h:40)
   100% ( 15 / 15) in 'Data.Vault.Lazy'
    80% ( 12 / 15) in 'Data.Vault.ST.Strict'
    Missing documentation for:
      Vault (src/Data/Vault/ST/backends/GHC.h:18)
      Key (src/Data/Vault/ST/backends/GHC.h:19)
      Locker (src/Data/Vault/ST/backends/GHC.h:40)
   100% ( 15 / 15) in 'Data.Vault.Strict'
  haddock: internal error: couldn't find ast for src/Data/Vault/ST/Lazy.hs ["src/Data/Vault/ST/ST.h","src/Data/Vault/ST/backends/GHC.h"]
  CallStack (from HasCallStack):
    error, called at utils/haddock/haddock-api/src/Haddock/Backends/Hyperlinker.hs:78:28 in main:Haddock.Backends.Hyperlinker
  
  --  While building package vault-0.3.1.3 using:
        /home/rgscott/.stack/setup-exe-cache/x86_64-linux/Cabal-simple_mPHDZzAJ_3.0.0.0_ghc-8.8.1 --builddir=.stack-work/dist/x86_64-linux/Cabal-3.0.0.0 haddock --html --hoogle --html-location=../$pkg-$version/ --haddock-option=--hyperlinked-source --haddock-option=--quickjump
      Process exited with code: ExitFailure 1
  ```
  </details>
* `cabal`

  <details>
  
  ```
  $ cabal get vault-0.3.1.3 && cd vault-0.3.1.3                                                      
  Unpacking to vault-0.3.1.3/

  $ /opt/ghc/8.8.1/bin/haddock --version
  Haddock version 2.23.0, (c) Simon Marlow 2006
  Ported to use the GHC API by David Waern 2006-2008

  $ cabal haddock -w /opt/ghc/8.8.1/bin/ghc --haddock-option=--hyperlinked-source
  Resolving dependencies...
  Build profile: -w ghc-8.8.1 -O1
  In order, the following will be built (use -v for more details):
   - vault-0.3.1.3 (first run)
  Configuring vault-0.3.1.3...
  Preprocessing library for vault-0.3.1.3..
  Running Haddock on library for vault-0.3.1.3..
  Warning: The documentation for the following packages are not installed. No
  links will be generated to these packages: hashable-1.3.0.0,
  semigroups-0.19.1, unordered-containers-0.2.10.0
  Haddock coverage:
  
  src/Data/Unique/Really.hs:6:1: warning: [-Wunused-imports]
      The import of ‘Control.Applicative’ is redundant
        except perhaps to import instances from ‘Control.Applicative’
      To import instances alone, use: import Control.Applicative()
    |
  6 | import Control.Applicative ((<$>))
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   100% (  4 /  4) in 'Data.Unique.Really'
  
  src/Data/Vault/ST/ST.h:15:1: warning: [-Wunused-imports]
      The import of ‘Data.Monoid’ is redundant
        except perhaps to import instances from ‘Data.Monoid’
      To import instances alone, use: import Data.Monoid()
     |
  15 | import Data.Monoid (Monoid(..))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  src/Data/Vault/ST/ST.h:16:1: warning: [-Wunused-imports]
      The import of ‘Data.Semigroup’ is redundant
        except perhaps to import instances from ‘Data.Semigroup’
      To import instances alone, use: import Data.Semigroup()
     |
  16 | import Data.Semigroup (Semigroup((<>)))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  src/Data/Vault/ST/ST.h:18:1: warning: [-Wunused-imports]
      The import of ‘Control.Applicative’ is redundant
        except perhaps to import instances from ‘Control.Applicative’
      To import instances alone, use: import Control.Applicative()
     |
  18 | import Control.Applicative ((<$>))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    80% ( 12 / 15) in 'Data.Vault.ST.Lazy'
    Missing documentation for:
      Vault (src/Data/Vault/ST/backends/GHC.h:18)
      Key (src/Data/Vault/ST/backends/GHC.h:19)
      Locker (src/Data/Vault/ST/backends/GHC.h:40)
   100% ( 15 / 15) in 'Data.Vault.Lazy'
  
  src/Data/Vault/ST/ST.h:15:1: warning: [-Wunused-imports]
      The import of ‘Data.Monoid’ is redundant
        except perhaps to import instances from ‘Data.Monoid’
      To import instances alone, use: import Data.Monoid()
     |
  15 | import Data.Monoid (Monoid(..))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  src/Data/Vault/ST/ST.h:16:1: warning: [-Wunused-imports]
      The import of ‘Data.Semigroup’ is redundant
        except perhaps to import instances from ‘Data.Semigroup’
      To import instances alone, use: import Data.Semigroup()
     |
  16 | import Data.Semigroup (Semigroup((<>)))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  src/Data/Vault/ST/ST.h:18:1: warning: [-Wunused-imports]
      The import of ‘Control.Applicative’ is redundant
        except perhaps to import instances from ‘Control.Applicative’
      To import instances alone, use: import Control.Applicative()
     |
  18 | import Control.Applicative ((<$>))
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    80% ( 12 / 15) in 'Data.Vault.ST.Strict'
    Missing documentation for:
      Vault (src/Data/Vault/ST/backends/GHC.h:18)
      Key (src/Data/Vault/ST/backends/GHC.h:19)
      Locker (src/Data/Vault/ST/backends/GHC.h:40)
   100% ( 15 / 15) in 'Data.Vault.Strict'
  Warning: Data.Unique.Really: could not find link destinations for:
      Hashable hashWithSalt hash
  haddock: internal error: couldn't find ast for src/Data/Vault/ST/Lazy.hs ["src/Data/Vault/ST/ST.h","src/Data/Vault/ST/backends/GHC.h"]
  CallStack (from HasCallStack):
    error, called at utils/haddock/haddock-api/src/Haddock/Backends/Hyperlinker.hs:78:28 in main:Haddock.Backends.Hyperlinker
  cabal: Failed to build documentation for vault-0.3.1.3.
  ```
  </details>

This is a regression from Haddock 2.22.0, which does not error:

<details>

```
$ /opt/ghc/8.6.5/bin/haddock --version
Haddock version 2.22.0, (c) Simon Marlow 2006
Ported to use the GHC API by David Waern 2006-2008

$ cabal haddock -w /opt/ghc/8.6.5/bin/ghc --haddock-option=--hyperlinked-source
Resolving dependencies...
Build profile: -w ghc-8.6.5 -O1
In order, the following will be built (use -v for more details):
 - vault-0.3.1.3 (first run)
Configuring vault-0.3.1.3...
Preprocessing library for vault-0.3.1.3..
Running Haddock on library for vault-0.3.1.3..
Warning: The documentation for the following packages are not installed. No
links will be generated to these packages: hashable-1.3.0.0,
semigroups-0.19.1, unordered-containers-0.2.10.0
Haddock coverage:
 100% (  4 /  4) in 'Data.Unique.Really'
  80% ( 12 / 15) in 'Data.Vault.ST.Lazy'
  Missing documentation for:
    Vault (src/Data/Vault/ST/backends/GHC.h:18)
    Key (src/Data/Vault/ST/backends/GHC.h:19)
    Locker (src/Data/Vault/ST/backends/GHC.h:40)
 100% ( 15 / 15) in 'Data.Vault.Lazy'
  80% ( 12 / 15) in 'Data.Vault.ST.Strict'
  Missing documentation for:
    Vault (src/Data/Vault/ST/backends/GHC.h:18)
    Key (src/Data/Vault/ST/backends/GHC.h:19)
    Locker (src/Data/Vault/ST/backends/GHC.h:40)
 100% ( 15 / 15) in 'Data.Vault.Strict'
Warning: Data.Unique.Really: could not find link destinations for:
    Hashable hashWithSalt hash
Documentation created:
/home/rgscott/Documents/Hacking/Haskell/vault-0.3.1.3/dist-newstyle/build/x86_64-linux/ghc-8.6.5/vault-0.3.1.3/doc/html/vault/index.html
```
</details>

I'll try to come up with a more minimal test case shortly.

--- a/haddock-api/src/Haddock.hs
+++ b/haddock-api/src/Haddock.hs
@@ -432,7 +432,7 @@ render dflags flags sinceQual qual ifaces installedIfaces extSrcMap = do
   when (Flag_HyperlinkedSource `elem` flags && not (null ifaces)) $ do
     withTiming (pure dflags') "ppHyperlinkedSource" (const ()) $ do
       _ <- {-# SCC ppHyperlinkedSource #-}
-           ppHyperlinkedSource odir libDir opt_source_css pretty srcMap ifaces
+           ppHyperlinkedSource (verbosity flags) odir libDir opt_source_css pretty srcMap ifaces
       return ()
 
 

--- a/haddock-api/src/Haddock/Backends/Hyperlinker.hs
+++ b/haddock-api/src/Haddock/Backends/Hyperlinker.hs
@@ -18,12 +18,13 @@ import Data.Maybe
 import System.Directory
 import System.FilePath
 
-import HieTypes       ( HieFile(..), HieASTs(..) )
+import HieTypes       ( HieFile(..), HieASTs(..), HieAST(..), NodeInfo(..) )
 import HieBin         ( readHieFile, hie_file_result)
 import Data.Map as M
 import FastString     ( mkFastString )
 import Module         ( Module, moduleName )
 import NameCache      ( initNameCache )
+import SrcLoc         ( mkRealSrcLoc, realSrcLocSpan )
 import UniqSupply     ( mkSplitUniqSupply )
 
 
@@ -65,26 +66,37 @@ ppHyperlinkedModuleSource verbosity srcdir pretty srcs iface = case ifaceHieFile
                  <$> (readHieFile (initNameCache u []) hfp)
 
         -- Get the AST and tokens corresponding to the source file we want
-        let mast | M.size asts == 1 = snd <$> M.lookupMin asts
-                 | otherwise        = M.lookup (mkFastString file) asts
+        let fileFs = mkFastString file
+            mast | M.size asts == 1 = snd <$> M.lookupMin asts
+                 | otherwise        = M.lookup fileFs asts
+            ast = fromMaybe (emptyHieAst fileFs) mast
+            fullAst = recoverFullIfaceTypes df types ast
             tokens = parse df file rawSrc
 
+        -- Warn if we didn't find an AST, but there were still ASTs
+        if M.null asts
+          then pure ()
+          else out verbosity verbose $ unwords [ "couldn't find ast for"
+                                               , file, show (M.keys asts) ]
+
         -- Produce and write out the hyperlinked sources
-        case mast of
-          Just ast ->
-              let fullAst = recoverFullIfaceTypes df types ast
-              in writeUtf8File path . renderToString pretty . render' fullAst $ tokens
-          Nothing
-            | M.size asts == 0 -> return ()
-            | otherwise -> do
-                out verbosity verbose $ unwords [ "couldn't find ast for"
-                                                , file, show (M.keys asts) ]
-                return ()
+        writeUtf8File path . renderToString pretty . render' fullAst $ tokens
     Nothing -> return ()
   where
     df = ifaceDynFlags iface
     render' = render (Just srcCssFile) (Just highlightScript) srcs
     path = srcdir </> hypSrcModuleFile (ifaceMod iface)
+
+    emptyNodeInfo = NodeInfo
+      { nodeAnnotations = mempty
+      , nodeType = []
+      , nodeIdentifiers = mempty
+      }
+    emptyHieAst fileFs = Node
+      { nodeInfo = emptyNodeInfo
+      , nodeSpan = realSrcLocSpan (mkRealSrcLoc fileFs 1 0)
+      , nodeChildren = []
+      }
 
 -- | Name of CSS file in output directory.
 srcCssFile :: FilePath

--- a/haddock-api/src/Haddock/Backends/Hyperlinker.hs
+++ b/haddock-api/src/Haddock/Backends/Hyperlinker.hs
@@ -7,7 +7,7 @@ module Haddock.Backends.Hyperlinker
 
 
 import Haddock.Types
-import Haddock.Utils (writeUtf8File)
+import Haddock.Utils (writeUtf8File, out, verbose, Verbosity)
 import Haddock.Backends.Hyperlinker.Renderer
 import Haddock.Backends.Hyperlinker.Parser
 import Haddock.Backends.Hyperlinker.Types
@@ -32,27 +32,28 @@ import UniqSupply     ( mkSplitUniqSupply )
 -- Note that list of interfaces should also contain interfaces normally hidden
 -- when generating documentation. Otherwise this could lead to dead links in
 -- produced source.
-ppHyperlinkedSource :: FilePath -- ^ Output directory
+ppHyperlinkedSource :: Verbosity
+                    -> FilePath -- ^ Output directory
                     -> FilePath -- ^ Resource directory
                     -> Maybe FilePath -- ^ Custom CSS file path
                     -> Bool -- ^ Flag indicating whether to pretty-print HTML
                     -> M.Map Module SrcPath -- ^ Paths to sources
                     -> [Interface] -- ^ Interfaces for which we create source
                     -> IO ()
-ppHyperlinkedSource outdir libdir mstyle pretty srcs' ifaces = do
+ppHyperlinkedSource verbosity outdir libdir mstyle pretty srcs' ifaces = do
     createDirectoryIfMissing True srcdir
     let cssFile = fromMaybe (defaultCssFile libdir) mstyle
     copyFile cssFile $ srcdir </> srcCssFile
     copyFile (libdir </> "html" </> highlightScript) $
         srcdir </> highlightScript
-    mapM_ (ppHyperlinkedModuleSource srcdir pretty srcs) ifaces
+    mapM_ (ppHyperlinkedModuleSource verbosity srcdir pretty srcs) ifaces
   where
     srcdir = outdir </> hypSrcDir
     srcs = (srcs', M.mapKeys moduleName srcs')
 
 -- | Generate hyperlinked source for particular interface.
-ppHyperlinkedModuleSource :: FilePath -> Bool -> SrcMaps -> Interface -> IO ()
-ppHyperlinkedModuleSource srcdir pretty srcs iface = case ifaceHieFile iface of
+ppHyperlinkedModuleSource :: Verbosity -> FilePath -> Bool -> SrcMaps -> Interface -> IO ()
+ppHyperlinkedModuleSource verbosity srcdir pretty srcs iface = case ifaceHieFile iface of
     Just hfp -> do
         -- Parse the GHC-produced HIE file
         u <- mkSplitUniqSupply 'a'
@@ -75,8 +76,10 @@ ppHyperlinkedModuleSource srcdir pretty srcs iface = case ifaceHieFile iface of
               in writeUtf8File path . renderToString pretty . render' fullAst $ tokens
           Nothing
             | M.size asts == 0 -> return ()
-            | otherwise -> error $ unwords [ "couldn't find ast for"
-                                           , file, show (M.keys asts) ]
+            | otherwise -> do
+                out verbosity verbose $ unwords [ "couldn't find ast for"
+                                                , file, show (M.keys asts) ]
+                return ()
     Nothing -> return ()
   where
     df = ifaceDynFlags iface

--- a/haddock-api/src/Haddock/Backends/Hyperlinker/Utils.hs
+++ b/haddock-api/src/Haddock/Backends/Hyperlinker/Utils.hs
@@ -102,7 +102,7 @@ type PrintedType = String
 -- >       hieAst
 --
 -- However, this is very inefficient (both in time and space) because the
--- mutliple calls to 'recoverFullType' don't share intermediate results. This
+-- multiple calls to 'recoverFullType' don't share intermediate results. This
 -- function fixes that.
 recoverFullIfaceTypes
   :: DynFlags

--- a/haddock.cabal
+++ b/haddock.cabal
@@ -46,6 +46,7 @@ extra-source-files:
   html-test/src/*.hs
   html-test/ref/*.html
   hypsrc-test/src/*.hs
+  hypsrc-test/src/*.h
   hypsrc-test/ref/src/*.html
   latex-test/src/**/*.hs
   latex-test/ref/**/*.tex

--- a/hypsrc-test/ref/src/Bug1091.html
+++ b/hypsrc-test/ref/src/Bug1091.html
@@ -1,0 +1,34 @@
+<html xmlns="http://www.w3.org/1999/xhtml"
+><head
+  ><link rel="stylesheet" type="text/css" href="style.css"
+     /><script type="text/javascript" src="highlight.js"
+    ></script
+    ></head
+  ><body
+  ><pre
+    ><span class="hs-pragma"
+      >{-# LANGUAGE CPP #-}</span
+      ><span
+      >
+</span
+      ><span id="line-2"
+      ></span
+      ><span class="hs-keyword"
+      >module</span
+      ><span
+      > </span
+      ><span class="hs-identifier"
+      >Bug1091</span
+      ><span
+      > </span
+      ><span class="hs-keyword"
+      >where</span
+      ><span class="hs-cpp"
+      >
+
+#include &quot;Include1For1091.h&quot;
+</span
+      ></pre
+    ></body
+  ></html
+>

--- a/hypsrc-test/src/Bug1091.hs
+++ b/hypsrc-test/src/Bug1091.hs
@@ -1,0 +1,4 @@
+{-# LANGUAGE CPP #-}
+module Bug1091 where
+
+#include "Include1For1091.h"

--- a/hypsrc-test/src/Include1For1091.h
+++ b/hypsrc-test/src/Include1For1091.h
@@ -1,0 +1,6 @@
+/* Include1For1091.h */
+
+foo :: Int
+foo = 42
+
+#include "Include2For1091.h"

--- a/hypsrc-test/src/Include2For1091.h
+++ b/hypsrc-test/src/Include2For1091.h
@@ -1,0 +1,4 @@
+/* Include2For1091.h */
+
+bar :: Int
+bar = 27

